Usermanagement

michaelottofits opened the pull request:
user management for oidc logins is needed 

the following tasks are required

- creation of users authenticated via oidc
- creation and assignment of groups contained in the oidc role claim

Commit:
add checkuser  this add user is not exists  for this operation the modul need root / sudo rights.  for testing set PAM_USER variable for your user and call binary with sudo -E ./pam-exec-oauth2

Commit:
get next uid for user

Commit:
user create with check and group add with member added

Commit:
add  a switch into config for add user and groups

Commit:
Update Readme add keycloak configuration

Commit:
fix github error lint messages

Commit:
add group membership by roles in token
add check old user added by pam modul to delete this user after 60 days

Commit:
fix delete check

Commit:
update sample and readme

michaelottofits commented:
@Gerrit91 can pls check and merge this ?

Gerrit91 commented:
Added @majst01 to the project maintainers and assigned him instead.

majst01 commented on go.mod:
```diff
@@ -9,4 +9,5 @@ require (
 	google.golang.org/appengine v1.6.7 // indirect
 	gopkg.in/square/go-jose.v2 v2.6.0
 	gopkg.in/yaml.v2 v2.4.0
+	github.com/tredoe/osutil v1.0.5
```
https://pkg.go.dev/os/user#Lookup

michaelottofits replied:
used see last commit

majst01 commented on main.go:
```diff
@@ -23,39 +23,59 @@ package main
+	"strings"
+	"time"
 
+	"github.com/tredoe/osutil/user"
```
https://pkg.go.dev/os/user#Lookup

michaelottofits replied:
used stdlib fpr opertation see last commit

majst01 commented on main.go:
```diff
@@ -150,27 +195,184 @@ func readConfig(filename string) (*config, error) {
+
+// createUser this create user is not exsits
+func createUser(username string) {
+	_, err := user.LookupUser(username)
```
https://pkg.go.dev/os/user#Lookup is part of stdlib, please prefer these !

michaelottofits replied:
use stdlib right now see last commit

majst01 commented on main.go:
```diff
@@ -150,27 +195,184 @@ func readConfig(filename string) (*config, error) {
+
+// createUser this create user is not exsits
+func createGroup(role string, username string) {
+	_, err := user.LookupGroup(role)
```
https://pkg.go.dev/os/user#LookupGroup

michaelottofits replied:
the stdlib is now used here as well

majst01 commented on main.go:
```diff
@@ -150,27 +195,184 @@ func readConfig(filename string) (*config, error) {
+	layout := time.ANSIC
+	t, err := time.Parse(layout, currenttime)
+
+	if err != nil {
```
If parsing of time does not work, nil is returned, this is not good

michaelottofits replied:
if time parse go wrong i use the current time  to do nothing is better

majst01 commented on main.go:
```diff
@@ -150,27 +195,184 @@ func readConfig(filename string) (*config, error) {
+
+		gid, err := user.AddGroup(role, username)
+
+		if err != nil {
```
Error Handling, please return  the err to the caller

michaelottofits replied:
changed  this method return error to caller

majst01 commented on main.go:
```diff
@@ -150,27 +195,184 @@ func readConfig(filename string) (*config, error) {
+
+		err := user.AddUsersToGroup(role, username)
+
+		if err != nil {
```
error must be returned to the caller

michaelottofits replied:
changed  this method return error to caller

majst01 commented on main.go:
```diff
@@ -150,27 +195,184 @@ func readConfig(filename string) (*config, error) {
+	var Users []string
+	file, err := os.Open("/etc/passwd")
+
+	if err != nil {
```
error must be returned to the caller

michaelottofits replied:
changed  this method return error to caller

majst01 commented on main.go:
```diff
@@ -150,27 +195,184 @@ func readConfig(filename string) (*config, error) {
+	for _, u := range getAllUsers() {
+
+		currentuser, err := user.LookupUser(u)
+		if err != nil {
```
will panic with nilpointer if a error occurs, please continue the loop here

michaelottofits replied:
i added continue to this loop

majst01 reviewed (changes requested):
Mainly error handling and usage of external library for user lookups instead of using stdlib.

Commit:
add look up group and user from stdlib

Commit:
- fix error handling in main
- return error statements
- if time parse go wrong set current time
-  use stdlib for lookup user and group

Commit:
use error in caller on create group and membership

michaelottofits commented:
@majst01 pls check the last commits

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
 	for _, role := range claims.Roles {
+
+		if addGroup {
+			_, err := createGroup(role, username)
```
addGroup should either not error out on existing group, or return a ErrGroupExist and control flow reacts accordingly

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+		}
+
+		if addmembership {
+			_, err := addUserToGroup(role, username)
```
same as for createGroup

majst01 replied:
simply return with a proper error

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+			_, err := createGroup(role, username)
+
+			if err != nil {
+				log.Printf("group: %s created" + role)
```
wrong log entry, group was not created

majst01 replied:
simply return with a proper error

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+			_, err := addUserToGroup(role, username)
+
+			if err != nil {
+				log.Printf("membership: user %s to %s created", role, username)
```
wrong log entry, user was not added to group

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+		// if no user then add one
+		if currentuser == nil {
+
+			cmd := exec.Command("usr/sbin/useradd", "-m", "-s", "/bin/bash", "-c", app, username)
```
/usr/sbin, better use common.LookupPath("useradd") and let the OS take care of finding the binary.

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+			cmd := exec.Command("usr/sbin/useradd", "-m", "-s", "/bin/bash", "-c", app, username)
+			stdoutStderr, err := cmd.CombinedOutput()
+			if err != nil {
+				log.Printf("user already exists: %s ", err.Error())
```
if useradd returns with an error, it can not be for sure that the user already exists. Need a helper func to detect if user is already present, which was done before already. --> Error message must be wrong, better return fmt.Errorf("unable to add user: %w", err)

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+}
+
+// createUser this create user is not exsits
+func createUser(username string) {
```
return an error here

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+
+			log.Printf("%s", stdoutStderr)
+		} else {
+			log.Printf("user already exists: %s skip create", username)
```
return here

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+			log.Printf("user already exists: %s skip create", username)
+		}
+	} else {
+		log.Printf("%s", err)
```
return fmt.Errorf("error looking up user:%w, err)

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+// getLastLogin for a user
+func getLastLogin(username string) time.Time {
+
+	cmd := exec.Command("last", "-1", username)
```
os.LookupPath

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+	cmd := exec.Command("last", "-1", username)
+	stdoutStderr, err := cmd.CombinedOutput()
+	if err != nil {
+		log.Printf("No lastlogin: %s ", err.Error())
```
is this the right assumption ?

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+
+	s := string(stdoutStderr)
+	split := strings.Split(s, " ")
+	x := strings.TrimSpace(split[2] + " " + split[3] + " " + split[4] + " " + split[5] + " " + split[6] + " " + split[7])
```
unsure if this is a reliable way to detect the last logintime, in my machine this looks like:
```
⌂3.55 1d [stefan:~/dev/github.com/metal-stack/pam-exec-oauth2] usermanagement* ± last -n1 stefan
stefan   pts/25       tmux(5398).%40   Wed Sep 29 10:51   still logged in

```

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+	t, err := time.Parse(layout, currenttime)
+
+	if err != nil {
+		log.Fatal(err)
```
terminates the application

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+}
+
+// parseloginTime get time from pased last command
+func parseloginTime(currenttime string) time.Time {
```
no extra func for 3 lines of code with a single usage

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+			return gid, nil
+		}
+		// if the group exists, a group with the same destination is created
+		return -1, nil
```
wrong group ID returned

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+	reader := bufio.NewReader(file)
+
+	for {
+		line, err := reader.ReadString('\n')
```
error handling missing

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+		}
+
+	}
+	return Users, nil
```
lowercase please

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+// getAllUsers list all users from passwd
+func getAllUsers() ([]string, error) {
+
+	var Users []string
```
lowercase users

majst01 commented on main.go:
```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
+
+	currentUsers, err := getAllUsers()
+
+	if err != nil {
```
only deletes users on error !, simple return the error here

Commit:
change procedure for create user

Commit:
refactor error handling in create user , group and membership

Commit:
this solution may have to be discussed again later and dealt with in another branch. therefore commented out here.

Commit:
move not needed func to commentout

michaelottofits commented:
@majst01  i have edited the points with the error handling and commented on the process. i have commented out the deleting for this branch. we may need another discussion on how to do this.

Commit:
fix error messages on create user

Commit:
fix error messages an create group

Commit:
remove not needed code here

Commit:
remove not need lib

michaelottofits commented:
this request is no longer needed. the development is requested by the metal-stack owner.
